lib/AST/Module.cpp

```diff
@@ -1190,7 +1192,14 @@ void ModuleDecl::getImportedModules(SmallVectorImpl<ImportedModule> &modules,
 void
 SourceFile::getImportedModules(SmallVectorImpl<ModuleDecl::ImportedModule> &modules,
                                ModuleDecl::ImportFilter filter) const {
-  assert(ASTStage >= Parsed || Kind == SourceFileKind::SIL);
+  // FIXME: Ideally we should assert that the file has been name bound before
+  // calling this function. However unfortunately that can cause issues for
+  // overlays which can depend on a Clang submodule for the underlying framework
+  // they are overlaying, which causes us to attempt to load the overlay again.
+  // We need to find a way to ensure that an overlay dependency with the same
+  // name as the overlay always loads the underlying Clang module. We currently
+  // handle this for a direct import from the overlay, but not when it happens
+  // through other imports.
   assert(filter && "no imports requested?");
```

Comment:
Yikes. Do you have the callframes for the exact cycle on hand? We will need to untangle it to push requestification into this component.

Reply:
@CodaFi I don't have the exact cycle to hand, but it looks like this:

```
swift::performNameBinding
NameBinder::getModule
ClangImporter::Implementation::finishLoadingClangModule
namelookup::getAllImports
ImportCache::getImportSet
collectExports
ModuleDecl::getImportedModulesForLookup
SourceFile::getImportedModules
```

The overlay's dependency on a submodule for its underlying framework causes us to go looking for imports in the top-level module, which maps to the overlay, and around we go. 

I've been looking at ways of breaking this as a part of requestifying name binding. Ideally we'd always map a same-named overlay dependency to the underlying Clang module, no matter how far down in the import set we are.